patchwork-extensions-block/src/main/java/net/patchworkmc/mixin/extensions/block/blockentity/MixinWorldChunk.java

```diff
@@ -0,0 +1,156 @@
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockEntityProvider;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.BlockView;
+import net.minecraft.world.chunk.ChunkSection;
+import net.minecraft.world.chunk.WorldChunk;
+
+import net.patchworkmc.impl.extensions.block.BlockContext;
+import net.patchworkmc.impl.extensions.block.Signatures;
+
+@Mixin(WorldChunk.class)
+public abstract class MixinWorldChunk {
+	/**
+	 * @param blockState
+	 * @return the blockEntity created by IForgeBlock.createTileEntity(BlockState, World)
+	 * , null if the block(blockstate) does not have a BlockEntity
+	 */
+	@Unique
+	private BlockEntity patchwork_createBlockEntity(BlockState blockState) {
+		BlockEntity blockEntity = null;
+
+		if (BlockContext.hasBlockEntity(blockState)) {
+			WorldChunk me = (WorldChunk) (Object) this;
+			IForgeBlockState forgeBlockState = (IForgeBlockState) blockState;
+			return forgeBlockState.createTileEntity(me.getWorld());
+		}
+
+		return blockEntity;
+	}
+
+	////////////////////////
+	/// createBlockEntity()
+	////////////////////////
+	@Inject(method = "createBlockEntity", locals = LocalCapture.CAPTURE_FAILHARD, cancellable = true, at = @At(value = "INVOKE", target = Signatures.BlockState_getBlock))
+	private void patchwork_createBlockEntity(BlockPos blockPos, CallbackInfoReturnable<BlockEntity> info, BlockState blockState) {
+		info.setReturnValue(patchwork_createBlockEntity(blockState));
+		info.cancel();
+	}
+
+	////////////////////////
+	/// loadBlockEntity
+	////////////////////////
+	// Block block = this.getBlockState(pos).getBlock();
+	// if (block instanceof BlockEntityProvider) {
+	//     blockEntity3 = ((BlockEntityProvider)block).createBlockEntity(this.world);
+	private static final ThreadLocal<Object> loadBlockEntity_blockEntity = BlockContext.createContext();
+	@Redirect(method = "loadBlockEntity", at = @At(value = "INVOKE", target = Signatures.BlockState_getBlock, ordinal = 0))
+	private Block patchwork_loadBlockEntity_getBlock(BlockState blockState) {
+		BlockEntity blockEntity = patchwork_createBlockEntity(blockState);
+
+		if (blockEntity != null) {
+			BlockContext.setContext(loadBlockEntity_blockEntity, blockEntity);
+		}
+
+		return BlockContext.hasBlockEntityBlockMarker(blockEntity != null);
+	}
+
+	@Redirect(method = "loadBlockEntity", at = @At(value = "INVOKE", target = Signatures.BlockEntityProvider_createBlockEntity, ordinal = 0))
+	private BlockEntity patchwork_loadBlockEntity_createBlockEntity(BlockEntityProvider dummy, BlockView view) {
+		BlockEntity blockEntity = BlockContext.releaseContext(loadBlockEntity_blockEntity);
+		return blockEntity;
+	}
+
+	////////////////////////
+	/// setBlockState()
+	////////////////////////
+	private static final ThreadLocal<Object> loadBlockEntity_blockState2 = BlockContext.createContext();
```

Comment:
`@Unique`

Reply:
in the latest commit, added for all `private static final` fields inside mixin classes